Add a theme collection component for TS and support for Premium and Partner collections

BogdanUngureanu opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

This PR introduces a new ThemeCollection component that will group various types of themes (e.g. Premium, Partner). 

The implementation is currently behind a feature flag since the implementation is not complete: we have to properly style it and add an additional layout component that will display collections and banners, for example. 

Related to https://github.com/Automattic/dotcom-forge/issues/3959, https://github.com/Automattic/dotcom-forge/issues/3989

## Proposed Changes

* Add a generic component for Themeshowcase that displays the themes in a swiper carousel
* Add two components for Premium and Partner themes
* The implementation is behind a flag and not visible to users

## Testing Instructions
- Use the Calypso live link and go to /themes page in incognito.
- Append the feature flag `flags=themes/discovery-lots`
- Notice that we display three carousels:
- - Premium themes 
- - Partner themes 
- Ignore the unstyled carousel, which will be handled in a separate PR.
<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->
![Screenshot 2023-10-06 at 17 27 40](https://github.com/Automattic/wp-calypso/assets/995955/daf4f896-ebe6-4090-aaf7-ce7f1f699a18)


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

BogdanUngureanu committed:
WIP: Add a theme collection component for TS

BogdanUngureanu committed:
Fixed typo

BogdanUngureanu committed:
Pass the correct props and use ThemeBlock instead

BogdanUngureanu committed:
Merge branch 'trunk' into add/theme-collection-component

BogdanUngureanu committed:
Add the key prop to avoid React warnings and guard the ThemeCollection behind a feature flag.

BogdanUngureanu committed:
Fixed feature flag condition. The component is now visible only when the lots flag is active and the user is logged out.

BogdanUngureanu committed:
Refine the feature flags.

For now, lits needs to be false too in development.

dsas committed:
Add support for subheadings on theme collections

dsas committed:
Temp: Make individual theme cards bigger so it is a better representation of what we want

BogdanUngureanu committed:
Add PremiumThemesCollection component

dsas committed:
First pass of theme collections as carousels

Still more work to be done, some mentioned in code, others are things
like support for multiple carousels on a single page.

BogdanUngureanu committed:
Pass the props directly

BogdanUngureanu committed:
Fixed null errors and set the feature flag to 0

BogdanUngureanu commented on client/components/theme-collection/index.tsx:
```diff
@@ -0,0 +1,112 @@
+import 'swiper/css/keyboard';
+import 'swiper/css/mousewheel';
+
+interface ThemeCollectionProps {
```
This interface needs to be refactored. I've duplicated the props in the PremiumCollection for now.

We'll need a generic interface that this interface will extend. 

We can do it in a separate PR.

BogdanUngureanu commented on client/components/theme-collection/index.tsx:
```diff
@@ -0,0 +1,112 @@
+	}, [ themes ] );
+
+	if ( themes === null ) {
+		return <></>; // @todo Add LoadingPlaceholders instead.
```
We should mark themes as nullable array. When themes is null, we should display a LoadingPlaceholder to avoid a flickering screen.

BogdanUngureanu committed:
Merge branch 'trunk' into add/theme-collection-component

dsas committed:
Use existing version of swiper

BogdanUngureanu committed:
Merge remote-tracking branch 'origin/add/theme-collection-component' into add/theme-collection-component

BogdanUngureanu committed:
Refactored the Theme collection Types (and added PartnerThemes) and fixed the SSR issues with the SwiperJS library.

BogdanUngureanu committed:
Commit yarn.lock updated after I executed yarn install

BogdanUngureanu commented on client/components/theme-collection/index.tsx:
```diff
@@ -0,0 +1,126 @@
+		if ( themes !== null && themes.length > 0 ) {
+			const el = document.querySelector( '.swiper-container' ) as HTMLElement;
+			if ( el ) {
+				Promise.all( [
```
We have to import the swiper package dynamically because it doesn't offer a CommonJS version. Because of this, the SSR build will fail.

From a performance standpoint, I don't think it will cause any issues; we are already loading the ThemeCards dynamically anyway.

dsas replied:
This should be a code comment really

BogdanUngureanu replied:
Good point!

BogdanUngureanu committed:
Add comment explaining why we need to load the swiper carousel dynamically.

dsas commented on client/components/theme-collection/index.tsx:
```diff
@@ -0,0 +1,131 @@
+				<div className="theme-collection__list-wrapper swiper-wrapper">
+					{ themes.map( ( theme, index ) => (
+						<div
+							key={ `theme-collection-container-${ collectionSlug }-${ index }` }
```
`index` is just the array index, react keys should use constant indexes not numeric array indexes because the zeroth item on the first render, could be the first item on the second render and it would update the wrong element.

This is probably only going to be a problem in this case if someone nulls out the react state and then someone changes curation order and so it refetches things, or something.

But still, easy enough to use theme.slug or something to generate a key.

BogdanUngureanu replied:
Sure, I think we can change it. The numeric approach was taken from the existing implementation of the theme list component.

BogdanUngureanu committed:
Change the key format to use theme.id, which contains the theme slug.

dsas committed:
Allow multiple ThemeCollection on one page (#82467)

Previously only the first ThemeCollection on each page worked as a
carousel. With this change all of the ThemeCollections work as
carousels.

One behaviour which may be a little...surprising is that keyboard
navigation will scroll all ThemeCollection that are in the current
viewport, not just the one which has focus.

Copons commented on client/my-sites/themes/themes-selection.jsx:
```diff
@@ -265,10 +269,51 @@ class ThemesSelection extends Component {
 		const interlacedThemes = interlaceThemes( themes, wpOrgThemes, query.search, isLastPage );
 
+		const isActiveInLits = config.isEnabled( 'themes/discovery-lits' ) && this.props.isLoggedIn;
+		const isActiveInLots = config.isEnabled( 'themes/discovery-lots' ) && ! this.props.isLoggedIn;
```
Nitpick: I find that `isActiveIn...` does not describe what we are doing here, and IMHO it's a bit confusing with all the `lits` and `lots`. 🙂 

How about something like this:

```js
import { isEnabled } from '@automattic/calypso-config';

// ...

const {
  // ...
  isLoggedIn,
} = this.props;

// Consider other names like `isThemeCollectionEnabled` etc.
const isDiscoveryEnabled =
  ( isEnabled( 'themes/discovery-lits' && isLoggedIn ) ||
  ( isEnabled( 'themes/discovery-lots' && ! isLoggedIn );


```

BogdanUngureanu replied:
Sure! :)

Copons commented on client/components/theme-collection/index.tsx:
```diff
@@ -0,0 +1,133 @@
+				/**
+				 * We have to import the swiper package dynamically because it doesn't offer a CommonJS version. Because of this, the SSR build will fail.
+				 *
+				 * From a performance standpoint, I don't think it will cause any issues; we are already loading the ThemeCards dynamically anyway.
```
```suggestion
				 * From a performance standpoint, it should not cause any issues; we are already loading the ThemeCards dynamically anyway.
```

Copons commented on client/components/theme-collection/index.tsx:
```diff
@@ -0,0 +1,133 @@
+	const swiperInstance = useRef< SwiperType | null >( null );
+	const swiperContainerId = `swiper-container-${ collectionSlug }`;
+
+	useEffect( () => {
```
This is initializing Swiper at every render of the component.

BogdanUngureanu replied:
Wouldn't the Swiper initialize just once because we pass `themes` const and because we check it for null values? 🤔

Copons commented on client/components/theme-collection/create-map-state-to-props.ts:
```diff
@@ -0,0 +1,21 @@
+			siteId,
+		};
+	};
+};
```
Let's use hooks instead of `connect` (which is soft-deprecated) for selecting from the Redux state.

e.g.
```js
/* use-theme-collection.ts */

import { useSelector } from 'calypso/state';

export function useThemeCollection( query ) {
  const siteId = useSelector( getSelectedSiteId );
  const themes = useSelector(
    ( state ) => getThemesForQueryIgnoringPage( state, 'wpcom', query ) ?? []
  );

  return { siteId, themes };
}


/* premium-themes-collection.tsx */

import useThemeCollection from './use-theme-collection';

const query = {
  // ...
};

export default function PremiumThemesCollection( props ) {
  const { siteId, themes } = useThemeCollection( query );

  // ...
}
```

Copons commented on client/components/theme-collection/collection-types/partner-themes-collection.tsx:
```diff
@@ -0,0 +1,36 @@
+export default connect(
+	createWpcomThemeCollectionMapStateToProps( query ),
+	null
+)( localize( PartnerThemesCollection ) );
```
Besides replacing `connect` with a hook (as mentioned in another comment), we should also localize via hook!

```js
import { useTranslate } from 'i18n-calypso';

export default function FoobarCollection() {
  const translate = useTranslate();

  return <>{ translate( 'Foobar!' ) }</>;
}
```

⚠️ All these notes about Redux, `connect`, and hooks are relevant for all the collection-types.

Copons commented on client/components/theme-collection/create-map-state-to-props.ts:
```diff
@@ -0,0 +1,21 @@
+	return ( state: IAppState, ownProps: ThemesCollectionProps ) => {
+		const siteId = getSelectedSiteId( state ) as unknown as string;
+
+		const themes = ( getThemesForQueryIgnoringPage( state, 'wpcom', query ) as never[] ) ?? []; // We should change it to a nullable value so that we can use LoadingPlaceholders.
```
I have no idea what `never[]` is, but it sounds terrifying.

Copons commented on client/components/theme-collection/index.tsx:
```diff
@@ -0,0 +1,133 @@
+	isActive,
+	getPrice,
+	isInstalling,
+}: ThemeCollectionProps ): ReactElement {
```
Nitpick: let's try to keep props ordered alphabetically (in all component functions and calls).

Copons commented on client/components/theme-collection/index.tsx:
```diff
@@ -0,0 +1,133 @@
+import { Icon, chevronLeft, chevronRight } from '@wordpress/icons';
+import { ReactElement, useEffect, useRef } from 'react';
+import './style.scss';
+import { Swiper as SwiperType } from 'swiper/types';
```
There is a [Swiper React](https://swiperjs.com/react) set of components, which I can see in the Calypso node_modules (`wp-calypso/node_modules/swiper/swiper-react.mjs`).

What's the reason for using the vanilla JS version instead?

dsas replied:
We did use those originally! The swiper people [say they're probably going to remove them](https://swiperjs.com/react) and recommend migrating to the Swiper Element API. The GitHub issues had people struggling with various uses of the Swiper Element API so I switched to ye olde vanilla js.

Copons replied:
I guess not a biggie, although I liked the React-y version.

Their warning is super vague ("likely", "in the future", etc.), and I wonder whether we would be safe with it for the time being. 🤔 

Oh well, if this works, then never mind my personal preferences. 🙂

dsas replied:
I liked it too, and I may just have the wrong instincts given our approach to dependency updates etc.

Using the react version seems like storing up a problem for the future. Swiper seems to move faster than we do in terms of only supporting the latest stuff and dropping support for older stuff. The last thing we want is more work when we're reacting to a security problem in 2028 when we're jumping from 10.1 to 20.1

Copons reviewed (changes requested):
I think we are on the right direction.

I've noted a few changes I'd like to see (mostly around using modern React), and an open question about using Swiper React.

BogdanUngureanu committed:
Update client/components/theme-collection/index.tsx

Co-authored-by: Jacopo Tomasone <Copons@users.noreply.github.com>

BogdanUngureanu committed:
Use selectors instead of connect and make coding styling changes.

BogdanUngureanu committed:
Pass swiperContainerId as param to useEffect.

BogdanUngureanu committed:
Fix a Redux warning in the selector

BogdanUngureanu committed:
Use children instead of passing the data as props and use memo function to reduce then number of re-renders in the Premium and Partner Collections

Copons commented on client/components/theme-collection/index.tsx:
```diff
@@ -0,0 +1,110 @@
+			{ children }
+		</div>
+	);
+}
```
What's the reasoning for `CollectionListItem`?

For what I can see, the only significant bit is the `swiper-slide` CSS class needed by Swiper.
Everything else seems superfluous, and even `swiper-slide` technically is. 😄 

If you need to style a slide (e.g. to give them a width), you can do that with `.theme-collection__list-wrapper > *`.
It then becomes the responsibility of the consumer to ensure the ThemeCollection children have an appropriate/stylable wrapper div.

If you need a specific slide class for the Swiper initialization, why not let the consumer define it?

e.g.
```jsx
// consumer
<ThemeCollection slideClass="foobar__slide" />

// Collection
function ThemeCollection( { slideClass } ) {
  // ...
  new Swiper( el, {
    // all the settings
    slideClass
  } );
  // ...
}
```

This would also give us the opportunity to offer further slider customizations to various Collection consumers (not the case right now, but it's good to keep in mind for the future).

Copons replied:
As a counterpoint, I think there's value in offering a `ThemeCollectionItem` component, which might decrease the consumer's cognitive load.

I'd recommend keeping it as simple as possible, with only the essential bits for now and maybe extending it in the future if needed. E.g.

```jsx
export default function ThemeCollectionItem( { children } ) {
  return (
    <div className="theme-collection__item swiper-slide">
      { children }
    </div>
  );
}
```

Copons commented on client/components/theme-collection/index.tsx:
```diff
@@ -0,0 +1,110 @@
+	);
+}
+
+export function CollectionListItem( {
```
If we want to keep this, let's:
- Rename it to `ThemeCollectionItem`.
- Move it to its own separate `theme-collection-item` file.

Copons commented on client/components/theme-collection/index.tsx:
```diff
@@ -0,0 +1,110 @@
+}: PropsWithChildren< { collectionSlug: string; themeId: string } > ) {
+	return (
+		<div
+			key={ `theme-collection-container-${ collectionSlug }-${ themeId }` }
```
`key` is only needed when the element is a child of a list.

It is not the case here; it is when the `Item` component is used (like you do in the two collection type components).

Copons commented on client/components/theme-collection/index.tsx:
```diff
@@ -0,0 +1,110 @@
+	return (
+		<div
+			key={ `theme-collection-container-${ collectionSlug }-${ themeId }` }
+			className="theme--collection__list-item swiper-slide"
```
```suggestion
			className="theme-collection__list-item swiper-slide"
```

(see [CSS guidelines](https://github.com/Automattic/wp-calypso/blob/7752207da247005f4b24898bad1807e16a474d92/docs/coding-guidelines/css.md))

Copons commented on client/components/theme-collection/collection-types/partner-themes-collection.tsx:
```diff
@@ -0,0 +1,49 @@
+import { useTranslate } from 'i18n-calypso';
+import { memo, ReactElement } from 'react';
+import * as React from 'react';
```
We shouldn't need this.

Copons commented on client/components/theme-collection/collection-types/premium-themes-collection.tsx:
```diff
@@ -0,0 +1,48 @@
+import { useTranslate } from 'i18n-calypso';
+import { memo, ReactElement } from 'react';
+import * as React from 'react';
```
We shouldn't need this.

BogdanUngureanu committed:
Update client/components/theme-collection/index.tsx

Co-authored-by: Jacopo Tomasone <Copons@users.noreply.github.com>

Copons commented on client/components/theme-collection/themes-collection-props.ts:
```diff
@@ -0,0 +1,8 @@
+	getScreenshotUrl: ( themeId: string ) => string;
+	isActive: () => boolean;
+	isInstalling: () => boolean;
+}
```
I think this is not needed anymore.

BogdanUngureanu replied:
No, I don't think so. :)

Copons commented on client/components/theme-collection/index.tsx:
```diff
@@ -0,0 +1,110 @@
+export default function ThemeCollection( {
+	collectionSlug,
+	heading,
+	subheading,
```
How do you feel about renaming these into `title` and `description`? IMHO it's more clear at first glance.

BogdanUngureanu replied:
sure :)

BogdanUngureanu committed:
Rename the item component to ThemeCollectionItem

Copons commented on client/components/theme-collection/use-theme-collection.ts:
```diff
@@ -0,0 +1,19 @@
+	tier: string;
+	filter: string;
+	search: string;
+	collection: string;
```
Is `collection` actually a query property?

BogdanUngureanu replied:
yup

dsas commented on client/components/theme-collection/index.tsx:
```diff
@@ -0,0 +1,110 @@
+
+	const [ isSwiperLoaded, setSwiperLoaded ] = useState( false );
+	useEffect( () => {
+		if ( ! isSwiperLoaded ) {
```
I think this can get this component and swiper out of sync. If the swiper instance gets cleaned up (for whatever reason) then `isSwiperLoaded` will remain true and the swiper instance won't get recreated.

```suggestion
		if ( ! isSwiperLoaded || ! swiperInstance?.current || swiperInstance.current?.destroyed ) {
```

BogdanUngureanu committed:
Update client/components/theme-collection/index.tsx

Co-authored-by: Dean Sas <dean.sas@automattic.com>

Copons commented on client/my-sites/themes/themes-selection.jsx:
```diff
@@ -247,6 +250,43 @@ class ThemesSelection extends Component {
+	getCarouselBlock = ( collectionSlug, theme, index ) => {
+		return (
+			<ThemeBlock
+				collectionSlug={ collectionSlug }
```
I don't think we need this.

Copons commented on client/my-sites/themes/themes-selection.jsx:
```diff
@@ -247,6 +250,43 @@ class ThemesSelection extends Component {
+				getPrice={ this.getPremiumThemePrice }
+				getScreenshotUrl={ this.getScreenshotUrl }
+				index={ index }
+				isActive={ this.isThemeActive }
```
Eventually, we will not show the active theme in both collections and regular lists.

We can filter out the active theme from `themes` and remove this prop from the collection's items.

BogdanUngureanu committed:
Remove unnecessary props

Copons commented on client/my-sites/themes/themes-selection.jsx:
```diff
@@ -247,6 +250,43 @@ class ThemesSelection extends Component {
+
+	isInstallingTheme = ( theme ) => {
+		return this.props.isInstallingTheme( theme );
+	};
```
I'm missing something here.

Why are we wrapping these functions instead of passing the props through?

E.g. can't we do this?
```jsx
<ThemeBlock
  getActionLabel={ this.props.getActionLabel( theme ) }
</ThemeBlock>
```

BogdanUngureanu replied:
I had it like that but it triggers a re-render when the props from ThemeSelection are called.

Copons commented on client/my-sites/themes/themes-selection.jsx:
```diff
@@ -261,14 +301,27 @@ class ThemesSelection extends Component {
+				{ isDiscoveryEnabled && (
+					<>
+						<PremiumThemesCollection>{ this.getCarouselBlock }</PremiumThemesCollection>
+						<PartnerThemesCollection>{ this.getCarouselBlock }</PartnerThemesCollection>
```
I think this is super confusing and overcomplicated.

Throwing an idea into the mix: collection-specific **information** instead of collection-specific **components**.

Hopefully the example could clarify what I'm thinking:

_(Note: this is all untested!)_
```jsx
// theme-collection/constants.jsx

export const THEME_COLLECTIONS = {
  premium: {
    query: { /* ... */ },
    heading: translate( 'Premium Themes' ),
    subheading: <p>{ translate( 'Lorem ipsum dolor sit amet' ) }</p>,
  }
}
```

```jsx
// themes-selection.jsx

getThemeCollection = ( slug ) => {
  const {
    query,
    heading,
    subheading,
  } = THEME_COLLECTIONS[ slug ];
  
  const { themes } = useThemeCollection( query );

  return (
    <>
      <QueryThemes query={ query } siteId="wpcom" />
      <ThemeCollection
        slug={ slug }
        heading={ heading }
        subheading={ subheading }
      >
        { themes && themes.map( ( theme, index ) => (
          <ThemeCollectionItem key={ theme.id }>
            <ThemeBlock /* all the props */ />
          </ThemeCollectionItem>
        ) ) }
      </ThemeCollection>
    </>
  );
}

render() {
  return(
    <>
      { isDiscoveryEnabled &&
        getThemeCollection( 'premium' )
      }
    </>
  );
}
```

Off the top of my head, this has 2 big advantages:
- Simplifies the hell out of everything, reducing the prop drilling.
- Does not couple any theme fetching logic with collections.
  - By and large, we're going to show this in 1 or 2 places (TS and DP). I haven't checked, but I presume that DP has a completely different theme fetching logic (IIRC it uses a different endpoint even).
  - If we create a collection-specific component using `QueryThemes`, we are fundamentally limiting it to the TS, which would remove its usefulness altogether (what's the point of such a specific component if we are only going to use it once?).

BogdanUngureanu replied:
> THEME_COLLECTIONS.premium.query

I have some mixed feelings about keeping logic code in a constants file, separated from the actual code. I'm fine with keeping the texts (e.g. heading), but not sure about the query.

It's interesting. However, there's one problem with this approach. The code would add up and will increase even more the complexity of the ThemeSelection.

An alternative is to have a new component called `ThemeCollectionsLayout` that would:
- define the collections (query, headings, etc).
- construct the layout (collections, upgrade banners)

Keep in mind that besides the collection components, we still need to add the routing logic. It will add up. :)

dsas replied:
> in a constants file

You could just call the file "collection-definitions.ts" if that helps

Copons replied:
> I have some mixed feelings about keeping logic code in a constants file

That really isn't logic though. It's the options argument for the themes endpoint. It would be the same as adding, like, `themesPerPage: 10`.

BogdanUngureanu committed:
Fix css class

dsas commented on client/components/theme-collection/collection-types/premium-themes-collection.tsx:
```diff
@@ -32,13 +33,13 @@ function PremiumThemesCollection( {
-						<CollectionListItem
+						<ThemeCollectionItem
 							key={ theme.id }
 							collectionSlug="premium-themes"
```
`premium-themes` could be moved up into a const rather than repeated.

BogdanUngureanu committed:
Refactor the implementation

BogdanUngureanu committed:
Rename props

BogdanUngureanu committed:
Remove collectionSlug from ThemeBlock.

BogdanUngureanu committed:
Remove unused props script.

BogdanUngureanu committed:
Remove unused code.

BogdanUngureanu committed:
Remove empty line.

Copons commented on client/my-sites/themes/theme-collections-layout.tsx:
```diff
@@ -0,0 +1,76 @@
+		title: string;
+		description: ReactElement;
+		query: ThemesQuery;
+	} ): ReactElement => {
```
Let's turn these types into a `ShowcaseThemeCollectionProps` interface for the sake of my eyes. 😆

Copons commented on client/my-sites/themes/theme-collections-layout.tsx:
```diff
@@ -0,0 +1,76 @@
+	return (
+		<>
+			<ShowcaseThemeCollection
+				key={ THEME_COLLECTIONS.premium.collectionSlug }
```
We don't need keys here, it's not a list, it's just two instances of the same component.

Copons commented on client/my-sites/themes/theme-collections-layout.tsx:
```diff
@@ -0,0 +1,76 @@
+	getOptions,
+	getScreenshotUrl,
+}: ThemesCollectionsLayoutProps ) {
+	const ShowcaseThemeCollection = ( {
```
I'm not convinced we should declare a component inside a component.
I presume the intent is to remove one level of props drilling, but it then becomes an antipattern and performance concern.
If we want to streamline this, we could do something like this instead:

```jsx
function ShowcaseThemeCollection( {
	collectionSlug,
	title,
	description,
	query,
	getActionLabel,
	getOptions,
	getScreenshotUrl,
} ) {
	// ...
}

function ThemeCollectionsLayout( props ) {
	return (
		<>
			<ShowcaseThemeCollection
				{ ...THEME_COLLECTIONS.premium }
				{ ...props }
			/>
		</>
	);
}
```

Copons commented on client/components/theme-collection/index.tsx:
```diff
@@ -0,0 +1,95 @@
+	return (
+		<div className="theme-collection__container ">
+			<h2>{ title }</h2>
+			{ description }
```
I'd recommend adding some specificity. I'm sure it will turn out to be useful. 🙂
```suggestion
			<h2 className="theme-collection__title">{ title }</h2>
			<div className="theme-collection__description">{ description }<div>
```

dsas replied:
100% will, I already did the same in https://github.com/Automattic/wp-calypso/pull/82519/files

Copons commented on client/my-sites/themes/theme-collections-layout.tsx:
```diff
@@ -0,0 +1,76 @@
+import { THEME_COLLECTIONS } from 'calypso/my-sites/themes/collection-definitions';
+import { ThemesQuery, useThemeCollection } from 'calypso/my-sites/themes/use-theme-collection';
+
+interface ThemesCollectionsLayoutProps {
```
```suggestion
interface ThemeCollectionsLayoutProps {
```

Copons commented on client/my-sites/themes/collection-definitions.tsx:
```diff
@@ -0,0 +1,30 @@
+		},
+		title: translate( 'Premium Themes' ),
+		collectionSlug: 'premium-themes',
+		description: <p>{ translate( 'Lorem ipsum dolor sit amet' ) }</p>,
```
Let's not translate placeholder text please!

BogdanUngureanu committed:
Extract ShowcaseThemeCollection in outside of the layout component and move the collection specific code into a separate directory.

dsas commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -402,26 +403,62 @@ class ThemeShowcase extends Component {
 		switch ( tabKey ) {
 			case staticFilters.MYTHEMES?.key:
 				return <ThemesSelection { ...themeProps } />;
+			case staticFilters.RECOMMENDED.key:
```
At some point, maybe when we drop the feature flag, maybe we should think about renaming RECOMMENDED. Don't think we need to change now though.

dsas commented:
Tracks check:
 * `calypso_upgrade_nudge_impression` (with `{ cta_name : theme-upsell }` )  is fired for each theme
   * Including those that are off-screen due to the carousel. Given live lots triggers the same event for themes that are off-screen due to scroll, maybe this is fine.
 * Scrolling the carousel whether by button, mousewheel etc doesn't trigger any tracks events, even when reaching the end. Live lots doesn't trigger an equivalent event
 * ❗Clicking on a theme in the carousel doesn't trigger a `calypso_themeshowcase_theme_click` event, unlike live lots

I don't think this needs changing in this PR, we already have a task later in the proejct for checking and creating tracks events, but thought it was worth being aware.

Copons reviewed (approved):
💥 LGTM 💥 

Thank you Bogdan for this PR's rollercoaster!

@dsas would you mind creating a project issue for the Tracks event you noted in the comment above? https://github.com/Automattic/wp-calypso/pull/82254#issuecomment-1750203937

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/9255148).
